jscomp/others/belt_Option.ml

```diff
@@ -0,0 +1,58 @@
+(* Copyright (C) 2017 Authors of BuckleScript
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * In addition to the permissions granted to you by the LGPL, you may combine
+ * or link a "work that uses the Library" with a publicly distributed version
+ * of this file to produce a combined library or application, then distribute
+ * that combined work under the terms of your choosing, with no requirement
+ * to comply with the obligations normally placed on you by section 4 of the
+ * LGPL version 3 (or the corresponding section of a later version of the LGPL
+ * should you choose to use a later version).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
+
+
+let getExn = function
+  | Some x -> x
+  | None -> [%assert "getExn"]
+
+let foldU opt default f = match opt with
+  | Some x -> (f x [@bs])
+  | None -> default
+
+let fold opt default f = foldU opt default (fun[@bs] x -> f x)
+
+let mapU opt f = match opt with
+  | Some x -> Some (f x [@bs])
+  | None -> None
+
+let map opt f = mapU opt (fun[@bs] x -> f x)
+
+let flatMapU opt f = match opt with
+  | Some x -> (f x [@bs])
+  | None -> None
+
+let flatMap opt f = flatMapU opt (fun[@bs] x -> f x)
+
+let getOrElse opt default = match opt with
+  | Some x -> x
+  | None -> default
+
+let has = function
+  | Some _ -> true
```

Comment:
How about naming these `isSome` and `isNone` like Rust? I think they convey what the function actually does better than `has`/`isEmpty`.

Reply:
agreed. In addition to `isSome` `isNone`:

`Option.unwrap(Some(A))` vs `Option.get(Some(A))`
`Option.unwrapOr(Some(A), B)` vs `Option.getOrElse(Some(A), B)` vs `Option.getWithDefault(Some(A), B)`

Reply:
`isSome` and `isNone` are nice, but @bobzhang pointed out that `x == None` is more performant so they actually might not be needed.

I'd say that we should keep `getExn` and `getWithDefault` as is for consistency with other Belt modules which have `get`, `getExn`, `getBy` etc.

Reply:
I agree with suffix consistency there but `Option.get` doesn't communicate to me that we are unwrapping / moving / extracting the value contained in the `Option`. `unwrapExn` `unwrapWithDefault` imo are more clear, but I don't feel super strongly about it